File: web/apps/app/lib/core/__tests__/model-identity-contract.test.ts
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { describe, expect, it } from 'vitest';
import type { AIModel, Provider } from '@oriveo/shared';
import { findModelInProvider } from '../provider-model-ops';
import { deduplicateByCanonical } from '../providers/catalog-model';

interface ContractModel {
  id: string;
  name: string;
  canonicalModelId?: string;
}

interface LookupCase {
  id: string;
  providerKind: Provider['kind'];
  query: string;
  expectedModelId: string | null;
  models: ContractModel[];
}

interface DedupeCase {
  id: string;
  models: ContractModel[];
  expectedModelIds: string[];
}

interface ContractFile {
  version: number;
  lookupCases: LookupCase[];
  dedupeCases: DedupeCase[];
}

const contractPath = path.resolve(
  process.cwd(),
  '../../../shared/model-contracts/model_identity_contract.v1.json',
);
const contract = JSON.parse(readFileSync(contractPath, 'utf8')) as ContractFile;

function makeModel(model: ContractModel): AIModel {
  return {
    id: model.id,
    name: model.name,
    canonicalModelId: model.canonicalModelId,
    capabilities: ['text'],
    reasoningModeAvailable: false,
    isAvailable: true,
    isDefault: false,
    priceTier: '',
  };
}

function makeProvider(providerKind: Provider['kind'], models: ContractModel[]): Provider {
  const catalogModels = models.map(makeModel);
  return {
    id: 'provider-contract',
    kind: providerKind,
    status: { kind: 'connected' },
    models: catalogModels,
    catalogModels,
    apiKey: '',
    apiKeyPreview: '',
  };
}

describe('model identity contract', () => {
  for (const testCase of contract.lookupCases) {
    it(`lookup: ${testCase.id}`, () => {
      const provider = makeProvider(testCase.providerKind, testCase.models);
      const resolved = findModelInProvider(provider, testCase.query);
      expect(resolved?.id ?? null).toBe(testCase.expectedModelId);
    });
  }

  for (const testCase of contract.dedupeCases) {
    it(`dedupe: ${testCase.id}`, () => {
      const deduped = deduplicateByCanonical(testCase.models.map(makeModel));
      expect(deduped.map((model) => model.id)).toEqual(testCase.expectedModelIds);
    });
  }
});

File: android/app/src/test/java/ai/oriveo/community/core/provider/ModelIdentityContractTest.kt
package ai.oriveo.community.core.provider

import ai.oriveo.community.core.model.AIModel
import ai.oriveo.community.core.model.ModelCapability
import ai.oriveo.community.core.model.Provider
import ai.oriveo.community.core.model.ProviderKind
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

class ModelIdentityContractTest {

    @Serializable
    private data class ContractFile(
        val version: Int,
        val lookupCases: List<LookupCase> = emptyList(),
        val dedupeCases: List<DedupeCase> = emptyList(),
        val storedIdentifierCases: List<StoredIdentifierCase> = emptyList(),
        val sameRemoteCases: List<SameRemoteCase> = emptyList(),
    )

    @Serializable
    private data class LookupCase(
        val id: String,
        val providerKind: String,
        val query: String,
        val expectedCanonicalModelId: String? = null,
        val expectedModelId: String? = null,
        val models: List<ContractModel> = emptyList(),
    )

    @Serializable
    private data class DedupeCase(
        val id: String,
        val models: List<ContractModel> = emptyList(),
        val expectedModelIds: List<String> = emptyList(),
    )

    @Serializable
    private data class StoredIdentifierCase(
        val id: String,
        val providerKind: String,
        val model: ContractModel,
        val expectedStoredModelId: String,
    )

    @Serializable
    private data class SameRemoteCase(
        val id: String,
        val providerKind: String,
        val left: ContractModel,
        val right: ContractModel,
        val expected: Boolean,
    )

    @Serializable
    private data class ContractModel(
        val id: String,
        val name: String,
        val canonicalModelId: String? = null,
    )

    private val json = Json { ignoreUnknownKeys = true }

    private val contract: ContractFile by lazy {
        val path = contractCandidates()
            .firstOrNull { Files.exists(it) }
            ?.normalize()
            ?: error("model identity contract fixture not found")
        json.decodeFromString(path.toFile().readText())
    }

    @Test
    fun `contract fixture loads`() {
        assertEquals(1, contract.version)
        assertFalse(contract.lookupCases.isEmpty())
    }

    @Test
    fun `lookup cases stay aligned`() {
        contract.lookupCases.forEach { testCase ->
            val kind = providerKind(testCase.providerKind)
            val provider = makeProvider(kind, testCase.models)
            val resolved = ModelSelectionUtils.matchingModel(provider.allModels, testCase.query)

            assertEquals(
                "lookup ${testCase.id}",
                testCase.expectedModelId,
                resolved?.id,
            )
            assertEquals(
                "canonical ${testCase.id}",
                testCase.expectedCanonicalModelId,
                resolved?.canonicalModelId ?: resolved?.id,
            )
        }
    }

    @Test
    fun `dedupe cases stay aligned`() {
        contract.dedupeCases.forEach { testCase ->
            val deduped = ModelSelectionUtils.deduplicateByCanonical(testCase.models.map(::makeModel))
            assertEquals("dedupe ${testCase.id}", testCase.expectedModelIds, deduped.map { it.id })
        }
    }

    @Test
    fun `stored identifier cases stay aligned`() {
        contract.storedIdentifierCases.forEach { testCase ->
            val stored = ModelSelectionUtils.preferredStoredModelIdentifier(makeModel(testCase.model))
            assertEquals("stored ${testCase.id}", testCase.expectedStoredModelId, stored)
        }
    }

    @Test
    fun `same remote cases stay aligned`() {
        contract.sameRemoteCases.forEach { testCase ->
            val result = ModelSelectionUtils.modelsShareSameRemoteModel(
                makeModel(testCase.left),
                makeModel(testCase.right),
                providerKind(testCase.providerKind),
            )
            assertEquals("same remote ${testCase.id}", testCase.expected, result)
        }
    }

    private fun providerKind(raw: String): ProviderKind =
        ProviderKind.fromRawValue(raw) ?: error("unknown provider kind: $raw")

    private fun makeModel(model: ContractModel): AIModel =
        AIModel(
            id = model.id,
            name = model.name,
            canonicalModelId = model.canonicalModelId,
            capabilities = listOf(ModelCapability.Text),
            isAvailable = true,
        )

    private fun makeProvider(kind: ProviderKind, models: List<ContractModel>): Provider {
        val resolved = models.map(::makeModel)
        return Provider(
            id = "provider-contract",
            kind = kind,
            models = resolved,
            catalogModels = resolved,
        )
    }

    private fun contractCandidates(): List<Path> {
        val cwd = Paths.get("").toAbsolutePath().normalize()
        return generateSequence(cwd) { current -> current.parent }
            .take(8)
            .map { current -> current.resolve("shared/model-contracts/model_identity_contract.v1.json") }
            .toList()
    }
}
